pinot-core/src/main/java/com/linkedin/pinot/core/operator/aggregation/groupby/DefaultGroupKeyGenerator.java

```diff
@@ -24,307 +24,480 @@
-   * dictionary id's.
-   *
-   * @param rawKey
-   * @return
+   * {@inheritDoc}
    */
-  private String rawKeyToStringGroupKey(long rawKey) {
-    // Decode the rawKey.
-    int cardinality = _cardinalities[0];
-    StringBuilder builder = new StringBuilder(_dictionaries[0].get((int) (rawKey % cardinality)).toString());
-    rawKey /= cardinality;
-    for (int i = 1; i < _numGroupByColumns; i++) {
-      builder.append(GroupByConstants.GroupByDelimiter.groupByMultiDelimeter);
-      cardinality = _cardinalities[i];
-      builder.append(_dictionaries[i].get((int) (rawKey % cardinality)));
-      rawKey /= cardinality;
-    }
-    return builder.toString();
+  @Override
+  public boolean hasMultiValueGroupByColumn() {
+    return _hasMultiValueGroupByColumn;
   }
 
+
   /**
-   * Returns an iterator of group by key (dictionary based) and the
-   * corresponding string group by key based on the actual column values.
-   * @return
+   * {@inheritDoc}
    */
   @Override
   public Iterator<GroupKey> getUniqueGroupKeys() {
-    if (_storageType == StorageType.ARRAY_BASED) {
-      return new ArrayBasedGroupKeyIterator();
-    } else {
-      final ObjectIterator<Map.Entry<Long, Integer>> iterator = _groupKeyToId.entrySet().iterator();
-      return new MapBasedGroupKeyIterator(iterator);
+    switch (_storageType) {
+      case ARRAY_BASED:
+        return new ArrayBasedGroupKeyIterator();
+      case LONG_MAP_BASED:
+        return new LongMapBasedGroupKeyIterator();
+      case ARRAY_MAP_BASED:
+        return new ArrayMapBasedGroupKeyIterator();
+      default:
+        throw new RuntimeException("Unsupported storage type.");
     }
   }
 
   /**
    * {@inheritDoc}
-   * @return
    */
   @Override
   public int getNumGroupKeys() {
-    return _numUniqueGroupKeys;
+    return _numGroupKeys;
   }
 
   /**
-   * Save the passed in group key:
-   * - For ARRAY_BASED storage, we store unique keys in a boolean array
-   *   (key is indexed into the array).
-   * - For MAP_BASED storage, we generate a unique new key, which is the total
-   *   number of unique keys, and use that as the group key. We also store the
-   *   mapping between the original group by key and the newly generated index.
+   * Helper function to get the group key associated to the long type raw key. If there is no one, generate a new group
+   * key for this raw key and put them into the map.
+   * (LONG_MAP_BASED storage type)
    *
-   * @param rawKey
-   * @return
+   * @param rawKey raw key.
+   * @return group key.
    */
   private int updateRawKeyToGroupKeyMapping(long rawKey) {
-    int groupKey;
-    if (_storageType == StorageType.ARRAY_BASED) {
-      int intRawKey = (int) rawKey;
-      if (!_uniqueGroupKeysFlag[intRawKey]) {
-        _uniqueGroupKeysFlag[intRawKey] = true;
-        _numUniqueGroupKeys++;
+    int groupKey = _groupKeyToId.get(rawKey);
```

Comment:
If this method is going to be called per doc, can you check the function cal overhead? Also please use Java 8 for the performance measurement.